File: ReadCore/src/main/java/com/readcore/adapters/persistence/JsonBookRepository.java
package com.readcore.adapters.persistence;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import com.readcore.domain.entities.Book;
import com.readcore.domain.repositories.BookRepository;

import java.io.*;
import java.lang.reflect.Type;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

/**
 * JSON-based implementation of BookRepository for offline storage
 */
public class JsonBookRepository implements BookRepository {
    private final String dataDirectory;
    private final Gson gson;
    private Map<String, Book> books;
    
    public JsonBookRepository(String dataDirectory) {
        this.dataDirectory = dataDirectory;
        this.gson = new GsonBuilder()
                .registerTypeAdapter(LocalDateTime.class, new LocalDateTimeAdapter())
                .setPrettyPrinting()
                .create();
        this.books = new HashMap<>();
        ensureDataDirectory();
        loadBooks();
    }
    
    private void ensureDataDirectory() {
        File dir = new File(dataDirectory);
        if (!dir.exists()) {
            dir.mkdirs();
        }
    }
    
    private File getBooksFile() {
        return new File(dataDirectory, "books.json");
    }
    
    private void loadBooks() {
        File file = getBooksFile();
        if (file.exists()) {
            try (Reader reader = new FileReader(file)) {
                Type type = new TypeToken<Map<String, Book>>(){}.getType();
                Map<String, Book> loaded = gson.fromJson(reader, type);
                if (loaded != null) {
                    books = loaded;
                }
            } catch (IOException e) {
                System.err.println("Error loading books: " + e.getMessage());
            }
        }
    }
    
    private void saveBooks() {
        try (Writer writer = new FileWriter(getBooksFile())) {
            gson.toJson(books, writer);
        } catch (IOException e) {
            System.err.println("Error saving books: " + e.getMessage());
        }
    }
    
    @Override
    public void save(Book book) {
        books.put(book.getId(), book);
        saveBooks();
    }
    
    @Override
    public Optional<Book> findById(String id) {
        return Optional.ofNullable(books.get(id));
    }
    
    @Override
    public List<Book> findAll() {
        return new ArrayList<>(books.values());
    }
    
    @Override
    public List<Book> searchByTitle(String title) {
        String searchTerm = title.toLowerCase();
        return books.values().stream()
                .filter(book -> book.getTitle().toLowerCase().contains(searchTerm))
                .collect(Collectors.toList());
    }
    
    @Override
    public List<Book> searchByAuthor(String author) {
        String searchTerm = author.toLowerCase();
        return books.values().stream()
                .filter(book -> book.getAuthor().toLowerCase().contains(searchTerm))
                .collect(Collectors.toList());
    }
    
    @Override
    public void delete(String id) {
        books.remove(id);
        saveBooks();
    }
    
    @Override
    public boolean exists(String id) {
        return books.containsKey(id);
    }
}

File: ReadCore/android/app/src/main/java/com/readcore/android/adapters/repositories/JsonBookRepository.kt
package com.readcore.android.adapters.repositories

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.readcore.android.domain.entities.Book
import com.readcore.android.domain.repositories.BookRepository
import kotlinx.coroutines.flow.first

private val Context.dataStore by preferencesDataStore(name = "books")

class JsonBookRepository(private val context: Context) : BookRepository {
    private val gson = Gson()
    private val booksKey = stringPreferencesKey("books_data")

    override suspend fun save(book: Book) {
        val books = findAll().toMutableList()
        books.removeIf { it.id == book.id }
        books.add(book)
        saveBooks(books)
    }

    override suspend fun findById(id: String): Book? {
        return findAll().find { it.id == id }
    }

    override suspend fun findAll(): List<Book> {
        val preferences = context.dataStore.data.first()
        val json = preferences[booksKey] ?: return emptyList()
        val type = object : TypeToken<List<Book>>() {}.type
        return gson.fromJson(json, type)
    }

    override suspend fun searchByTitle(title: String): List<Book> {
        return findAll().filter { it.title.contains(title, ignoreCase = true) }
    }

    override suspend fun searchByAuthor(author: String): List<Book> {
        return findAll().filter { it.author.contains(author, ignoreCase = true) }
    }

    override suspend fun delete(id: String) {
        val books = findAll().filter { it.id != id }
        saveBooks(books)
    }

    override suspend fun exists(id: String): Boolean {
        return findById(id) != null
    }

    private suspend fun saveBooks(books: List<Book>) {
        context.dataStore.edit { preferences ->
            preferences[booksKey] = gson.toJson(books)
        }
    }
}
